### PR TITLE
Connection refactoring

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: quinn-proto/src/connection/mod.rs [hunk 1]
```diff
@@ -432,21 +432,15 @@ where
         let mut congestion_blocked = false;
 
         // Iterate over all spaces and find data to send
-        let mut space_idx = 0;
-        let spaces = [SpaceId::Initial, SpaceId::Handshake, SpaceId::Data];
-        while space_idx < spaces.len() {
-            let space_id = spaces[space_idx];
-
+        for space_id in SpaceId::iter() {
```
Comment (Matthias247) [inline quinn-proto/src/connection/mod.rs:456]: This was explicitly changed for GSO, and changing it back won't work. There's a comment at the end of the function why.

Comment (djc) [reply quinn-proto/src/connection/mod.rs]: Ah, oops. The comment wasn't very discoverable from the point of where the iteration happens, so I replaced my previous commit with a new one that explains this closer to the loop setup (8fa5c32).

**Comment by Ralith** [comment]
There appear to be two commits with the description "quinn-proto: move datagram receive logic into DatagramState" which do different things. Disambiguate? Looks good otherwise.

**Comment by djc** [comment]
Added a single character to disambiguate `receive` logic from `received` logic -- is that enough? I couldn't really come up with a better description for the second commit. Also rebased on current main and tacked on a final commit that just moves all of the `bytes_source` module's contents into `connection::streams::send`.

**Comment by djc** [comment]
I've updated the "extract separate *Stream" commits to add `send_stream()`/`recv_stream()` accessors directly on `Connection` rather than going through `Streams`.

**Comment by Ralith** [comment]
> I couldn't really come up with a better description for the second commit

Maybe "incoming datagram frame handling" instead of "datagram received logic"?

**Comment by djc** [comment]
> Maybe "incoming datagram frame handling" instead of "datagram received logic"?

Done.

**Review by Ralith — APPROVED** [review]

**Pull request merged**